source/fuzz/fuzzer_pass_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,176 @@
+
+        // Width is specified in the first index for either OpTypeInt or
+        // OpTypeFloat.
+        uint32_t width = GetIRContext()
+                             ->get_def_use_mgr()
+                             ->GetDef(operand_type_id)
+                             ->GetSingleWordOperand(0);
+        // Get the scalar type.
+        auto type = GetIRContext()->get_type_mgr()->GetType(operand_type_id);
+        // Whether the constant is signed, not used for float type so default
+        // set to true.
+        bool is_signed_constant =
+            type->AsInteger() ? type->AsInteger()->IsSigned() : true;
+
+        // Populate components based on vector type and size.
+        for (uint32_t i = 0; i < component_count; ++i) {
+          if (i == position) {
+            vec1_components.emplace_back(target_id1);
+            vec2_components.emplace_back(target_id2);
+          } else {
+            if (type->AsInteger()) {
+              // Operands are integers. Add random integers to each vector.
+              int sign1 = is_signed_constant
+                              ? (GetFuzzerContext()->ChooseEven() ? 1 : -1)
+                              : 1;
+              int sign2 = is_signed_constant
+                              ? (GetFuzzerContext()->ChooseEven() ? 1 : -1)
+                              : 1;
+              int random_int1 =
+                  sign1 *
+                  (GetFuzzerContext()->GetRandomIntegerFromRange(1, 100));
+              int random_int2 =
+                  sign2 *
+                  (GetFuzzerContext()->GetRandomIntegerFromRange(1, 100));
+              vec1_components.emplace_back(FindOrCreateIntegerConstant(
+                  fuzzerutil::IntToWords(random_int1, width,
+                                         is_signed_constant),
+                  width, is_signed_constant, false));
+              vec2_components.emplace_back(FindOrCreateIntegerConstant(
+                  fuzzerutil::IntToWords(random_int2, width,
+                                         is_signed_constant),
+                  width, is_signed_constant, false));
+            } else if (type->AsFloat()) {
+              // Operands are floats. Add random floats to each vector.
+              float sign1 = GetFuzzerContext()->ChooseEven() ? 1.0f : -1.0f;
+              float sign2 = GetFuzzerContext()->ChooseEven() ? 1.0f : -1.0f;
+              float random_float1 =
+                  sign1 *
+                  GetFuzzerContext()->GetRandomFloatFromRange(0.1f, 10.0f);
+              float random_float2 =
+                  sign2 *
+                  GetFuzzerContext()->GetRandomFloatFromRange(0.1f, 10.0f);
+              vec1_components.emplace_back(FindOrCreateFloatConstant(
+                  {fuzzerutil::FloatToWord(random_float1)}, width, false));
+              vec2_components.emplace_back(FindOrCreateFloatConstant(
+                  {fuzzerutil::FloatToWord(random_float2)}, width, false));
```

Comment:
Please take a look at the following text in [our PLDI paper](https://dl.acm.org/doi/pdf/10.1145/3453483.3454092):

> As another example, AddStore(𝑐, 0, 𝑠,𝑖) in Figure 4 adds ... ChangeIrrelevantValue, could then be employed to change the use of 0 to something more interesting (such as the input 𝑖).

Rather than choosing an explicit random value in the fuzzer pass, a better approach is to set each component (different from `position`) to an "irrelevant zero". That is, use `FuzzerPass:FindOrCreateZeroConstant`, passing `true` as the `is_irrelevant` argument.

This means that whenever the transformation is applied, it will initially fill all components with zero. However, these zeros will have the "irrelevant value" fact, so that spirv-fuzz knows that it can later replace them with whatever it likes - arbitrarily more complex expressions.

Reply:
It might seem that a downside to this is that it prohibits using division (due to the use of zero). But actually the current design also has problems related to division by zero that I can tell you about offline.

Reply:
Done. Im curious though as to why the current design couldnt address division by zero.